docs/docs/adrs/adr-015-partial-set-security.md

```diff
@@ -0,0 +1,308 @@
+
+// returns the N if Top N chain, otherwise an error
+func (k Keeper) GetTopN(ctx sdk.Context, chainID string) (uint8, error)
+```
+
+In a future version of PSS, we intend to extend the `interchain-security-pd query provider list-consumer-chains` query to also return information on whether a consumer chain is an Opt In or a Top N chain and with what N. This way, block explorers like [Mintscan](https://www.mintscan.io/) would show for a chain “This chain is secured by N% of Cosmos Hub.”
+
+### How do validators opt in?
+A validator can opt in by sending a new type of message that we introduce in [tx.proto](https://github.com/cosmos/interchain-security/blob/v3.3.0/proto/interchain_security/ccv/provider/v1/tx.proto#L1).
+```
+message MsgOptIn {
+    // the chain id of the consumer chain to opt in to
+    string chain_id = 1;
+    // the provider address of the validator on Cosmos Hub
+    string provider_addr = 2;
+}
+```
+Note that in a Top N consumer chain, the top N% Cosmos Hub validators are forced by default to validate the consumer chain. Nevertheless, validators in the bottom 100 - N% can opt in to validate as well.
+Additionally, note that due to undelegations, etc. a validator that is in the top N% validators might fall down to the 100 - N% in which case this validator is not forced to validate anymore. Therefore, a validator that might move out of the top N% might want to send a `MsgOptIn` to keep validating even if this happens.
+
+Note that a validator can send a `MsgOptIn` message even if the consumer chain is not yet running. To do this we reuse the [`IsConsumerProposedOrRegistered`](https://github.com/cosmos/interchain-security/blob/v3.3.0/x/ccv/provider/keeper/key_assignment.go#L644). If the `chainID` does not exist, the `MsgOptIn` should fail, as well as if the provider address does not exist.
+
+Additionally, a validator that opts in can then use `MsgAssignConsumerKey` to change the consumer key on the consumer chain as is currently done in Replicated Security.
+
+#### State & Query
+We store the list of validators that have opted in and at the block height they opted it in under the key:
+```
+ChainIdWithLenKey(OptedInByteKey, chainID)
+```
+
+And for simplicity for each validator we store a boolean on whether this validator has opted in on this chain or not.
+```
+ChainIdAndConsAddrKey(OptedInByteKey, chainID, addr)
+```
+
+We augment the state of the provider to keep track for each chain all the validators that have opted in and if a specific validator has opted in.
+```
+// returns all the validators that have opted in on chain `chainID`
+func (k Keeper) GetOptedInValidators(ctx sdk.Context, chainID string) []Validators
+
+func (k Keeper) IsValidatorOptedIn(ctx sdk.Context, chainID string, val Validator) bool
+```
+
+We introduce a query to retrieve the validators that are opted in. The query would operate as follows.
+```
+gaiad query provider optedInValidators $chainId
+```
+
+#### When do validators opt in?
```

Comment:
To implement top-n, we should just have an automatic process every block or epoch or whatever, which automatically opts in validators in the top-n, just as if they had sent the message themselves.